### PR TITLE
Build an unprivileged container image

### PR DESCRIPTION
**Pull request by bashofmann**
This adds building an additional container image with the `-unprivileged` suffix. The resulting container is not running with a root user. Otherwise it is the same as the standard qdrant container image.

Having an image that is running as nonRoot is necessary in a lot of enterprise environments due to security and compliance restrictions.

I tested this locally like this:

* Build the container image with the new Dockerfile
* Run the container like this `docker run --rm -p 6333:6333 -v $(pwd)/qdrant_storage:/qdrant/storage:z qdrant/qdrant`
* Ensure that a collection can be created, vectors can be inserted and queried
* Ensure that the vectors can still be queried after a restart of the container

My test script

```
from qdrant_client import QdrantClient
from qdrant_client.http.models import Distance, VectorParams
from qdrant_client.http.models import PointStruct

client = QdrantClient("localhost", port=6333)

client.recreate_collection(
    collection_name="test_collection",
    vectors_config=VectorParams(size=4, distance=Distance.DOT),
)

operation_info = client.upsert(
    collection_name="test_collection",
    wait=True,
    points=[
        PointStruct(id=1, vector=[0.05, 0.61, 0.76, 0.74], payload={"city": "Berlin"}),
        PointStruct(id=2, vector=[0.19, 0.81, 0.75, 0.11], payload={"city": "London"}),
        PointStruct(id=3, vector=[0.36, 0.55, 0.47, 0.94], payload={"city": "Moscow"}),
        PointStruct(id=4, vector=[0.18, 0.01, 0.85, 0.80], payload={"city": "New York"}),
        PointStruct(id=5, vector=[0.24, 0.18, 0.22, 0.44], payload={"city": "Beijing"}),
        PointStruct(id=6, vector=[0.35, 0.08, 0.11, 0.44], payload={"city": "Mumbai"}),
    ]
)
print(operation_info)

search_result = client.search(
    collection_name="test_collection",
    query_vector=[0.2, 0.1, 0.9, 0.7],
    limit=3
)
print(search_result)
```

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### CODE REVIEW

**Comment by generall** [comment]
Maybe it is possible to reuse dockerfile and only switch user by buildarg?

**Comment by bashofmann** [comment]
Good idea, I'll try that.

**Comment by bashofmann** [comment]
Done.

Result:

```
❯ docker build -t qdrant/qdrant .
...
❯ docker run -it --rm --entrypoint id qdrant/qdrant
uid=0(root) gid=0(root) groups=0(root)

❯ docker build --build-arg='USER_UID=1000' --build-arg='USER_GID=1000' --build-arg='USERNAME=qdrant' -t qdrant/qdrant:unprivileged .
...
❯ docker run -it --rm --entrypoint id qdrant/qdrant:unprivileged
uid=1000 gid=1000 groups=1000
```

File: .github/workflows/docker-image.yml [hunk 1]
```diff
@@ -43,3 +43,11 @@ jobs:
...
         docker tag $DOCKERHUB_TAG $GITHUB_TAG
         docker push $GITHUB_TAG
+        
+        DOCKERHUB_TAG_UNPRIVILEGED="qdrant/qdrant:${{ github.ref_name }}-unprivileged"
+        TAGS_UNPRIVILEGED="-t ${DOCKERHUB_TAG_UNPRIVILEGED}"
+        GITHUB_TAG_UNPRIVILEGED="docker.pkg.github.com/qdrant/qdrant/qdrant:${{ github.ref_name }}-unprivileged"
+      
+        docker buildx build --build-arg='USER_UID=1000' --build-arg='USER_GID=1000' --build-arg='USERNAME=qdrant' --platform='linux/amd64,linux/arm64' $TAGS_UNPRIVILEGED --push --label "org.opencontainers.image.version"=$RELEASE_VERSION .
+        docker tag $DOCKERHUB_TAG_UNPRIVILEGED $GITHUB_TAG_UNPRIVILEGED
+        docker push $GITHUB_TAG_UNPRIVILEGED
```
Comment (timvisee) [inline .github/workflows/docker-image.yml:53]: Should we also tag `:latest-unprivileged`?

File: Dockerfile [hunk 2]
```diff
@@ -119,11 +122,17 @@ COPY --from=builder /static ${APP}/static
 
 WORKDIR ${APP}
 
+# Create the user
+RUN if [[ "${USER_UID}" != "0" ]]; then (groupadd --gid $USER_GID $USERNAME \
+    && useradd --uid $USER_UID --gid $USER_GID -m $USERNAME \
+    && chown -R $USER_GID:$USER_GID ${APP}); fi
```
Comment (timvisee) [inline Dockerfile:128]: When all is stuffed together anyway, maybe we can do this instead:

```suggestion
# Create the user
RUN [[ "${USER_UID}" != "0" ]] \
    && groupadd --gid $USER_GID $USERNAME \
    && useradd --uid $USER_UID --gid $USER_GID -m $USERNAME \
    && chown -R $USER_GID:$USER_GID ${APP}
```

Comment (bashofmann) [reply Dockerfile:128]: Wouldn't this fail the build for USER_UID=0 (root user) like this?

Comment (timvisee) [reply Dockerfile:128]: You're right. A `|| true` suffix could fix that. But that would be ugly.

File: .github/workflows/docker-image.yml (hunk 1 above)
Comment (bashofmann) [reply .github/workflows/docker-image.yml:53]: I don't need it for my use-case, but probably a good idea to stay consistent. @generall Thoughts?

File: Dockerfile [hunk 3]
```diff
@@ -109,6 +109,9 @@ RUN apt-get update \
     && rm -rf /var/lib/apt/lists/*
 
 ARG APP=/qdrant
+ARG USERNAME=root
+ARG USER_UID=0
+ARG USER_GID=0
```
Comment (ffuugoo) [inline Dockerfile:114]: I suggest moving this close to the point of first use. (E.g., right before `RUN` that creates the user.)
It provides better caching.

Also, I'd personally reduce this to a single `USER_ID` ARG and derive everything else from there. E.g., use `USER_ID` for both `uid` and `gid` and simply "hardcode" the username to `user` or `qdrant`.

File: Dockerfile (hunk 2 above)
Comment (ffuugoo) [inline Dockerfile:128]:
```suggestion
RUN if [[ "$USER_UID" == 0 ]]; then \
        groupadd --gid "$USER_UID" qdrant; \
        useradd --uid "$USER_UID" --gid "$USER_GID" qdrant; \
        chown -R "$USER_UID:$USER_GID" "$APP"; \
    fi
```

Comment (bashofmann) [reply Dockerfile]: Done

**Review by timvisee — APPROVED** [review]
Once we resolve the current open review comments; LGTM.

File: .github/workflows/docker-image.yml (hunk 1 above)
Comment (bashofmann) [reply .github/workflows/docker-image.yml]: I added `:latest-unprivileged`

File: Dockerfile (hunk 3 above)
Comment (bashofmann) [reply Dockerfile]: I simplified and re-ordered it as suggested.

**Review by ffuugoo — APPROVED** [review]

**Pull request merged**